### PR TITLE
Refactor Course Statistics Page Routing

### PR DESCRIPTION
**Pull request by bivanalhar**
## Background

Previously, the calling of the API for fetching all statistics data, for all 3 tabs (students, staff, course progression, course performance) were all combined within one component, `index.jsx` inside Course Statistics. This might be okay for now, but since we are planning to extend the statistics to include some other information (e.g. assessment), adding the API furthermore later will make this page very cluttered.

Also, intuitively, one should only load the necessary information that is to be loaded and displayed inside the page. Therefore, we attempted to refactor the Course Statistics Page here, to make it easier for us later on to extend the page

## What has been done

- Refactor most of the component inside Statistics Page into Typescript
- Subsequently, introduce selector and also proper typing (not using PropTypes anymore for most of the components)
- Re-route the sidebars for statistics into students statistics initially
- Implement StatisticBars so that user can navigate between students, staff, and course statistics flawlessly and smoothly

## To-Do

- [x] Migrate the table inside Students and Staff Statistics into using Sortable / Filterable Table Component
- [x] Migrate the StudentStatisticsTable, StudentPerformanceTable, StudentProgressionTable into Typescript
- [x] Afterwards, migrate all remaining PropTypes there into `types.ts` and remove that PropTypes completely

### CODE REVIEW

File: client/app/bundles/course/statistics/pages/StatisticsIndex/course/StudentPerformanceTable.tsx [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+    students.sort((s1, s2) => (s2.correctness ?? 0) - (s1.correctness ?? 0));
+  }
+
+  const studentsWithHighlight = students.map((student, index) => ({
+    ...student,
+    isTopStudent: index <= numHighlightedStudents,
+    isBottomStudent: index >= students.length - numHighlightedStudents,
+  }));
+
+  const getStudentHighlightColor = (student: CourseStudent): string => {
```
Comment (ekowidianto) [inline client/app/bundles/course/statistics/pages/StatisticsIndex/course/StudentPerformanceTable.tsx:136]: Move this out from the component so it doesn't get re-declared everytime the component is re-rendered.

File: client/app/api/course/Statistics/CourseStatistics.js [hunk 2]
```diff
@@ -42,6 +42,6 @@ export default class CourseStatisticsAPI extends BaseCourseAPI {
   }
 
   get #urlPrefix() {
-    return `/courses/${this.courseId}/statistics/course`;
+    return `/courses/${this.courseId}/statistics`;
```
Comment (ekowidianto) [inline client/app/api/course/Statistics/CourseStatistics.js:45]: Can you rewrite this into ts and properly add the type for the api response.

File: client/app/bundles/course/statistics/handles.ts [hunk 3]
```diff
@@ -16,12 +14,8 @@ export const videoWatchHistoryHandle: DataHandle = (match) => {
       const { data } = await CourseAPI.users.fetch(userId);
 
       return {
-        activePath: `/courses/${courseId}/statistics`,
+        activePath: `/courses/${courseId}/statistics/students`,
         content: [
-          {
-            title: StatisticsIndex.handle,
```
Comment (ekowidianto) [inline client/app/bundles/course/statistics/handles.ts:22]: Why is this removed?

File: client/app/bundles/course/statistics/pages/StatisticsIndex/staff/StaffStatistics.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import Preload from 'lib/components/wrappers/Preload';
+
+import StatisticsTabs from '../StatisticsTabs';
+
+import StaffStatisticsTable from './StaffStatisticsTable';
+
+const StaffStatistics: FC = () => {
+  return (
+    <>
+      <StatisticsTabs />
```
Comment (ekowidianto) [inline client/app/bundles/course/statistics/pages/StatisticsIndex/staff/StaffStatistics.tsx:14]: Why not create a component <StatisticsContainer /> which is the parent of Student, Staff and Course Stats and place it in the routes with the path: 'statistics'? See how the element in the `path: 'courses/:courseId'` is written

File: client/app/lib/constants/sharedConstants.ts [hunk 5]
```diff
@@ -78,3 +79,5 @@ export const DEFAULT_LOCALE = process.env.DEFAULT_LOCALE ?? 'en';
 
 export const DEFAULT_TIME_ZONE =
   process.env.DEFAULT_TIME_ZONE ?? 'Asia/Singapore';
+
+export const NUM_CELL_CLASS_NAME = 'text-right';
```
Comment (ekowidianto) [inline client/app/lib/constants/sharedConstants.ts:83]: You might want to just add this to the table cell component related file.

**Review by ekowidianto — COMMENTED** [review]
LGTm thanks some ocmments below @bivanalhar

File: client/app/bundles/course/statistics/handles.ts (hunk 3 above)
Comment (bivanalhar) [reply client/app/bundles/course/statistics/handles.ts:22]: this line seems like serving no purpose at all to the component here (this one is videoWatchHistoryHandle, which basically has no relation at all with the StatisticsIndex).

Also, removing this does not impact the related page as well (`video_submissions`), therefore I am wondering why this line should be here in the first place.

File: client/app/lib/constants/sharedConstants.ts (hunk 5 above)
Comment (bivanalhar) [reply client/app/lib/constants/sharedConstants.ts:83]: if I do that, there will be linter error of that text 'text-right' being repeated more than some number of times (if I remembered correctly, the threshold is 6)

I have also considered an option of putting this NUM_CELL_CLASS_NAME inside Table component, however I think the current one that is here is much cleaner, and also might be useful for any other purposes since this is sharedConstants..

File: client/app/bundles/course/statistics/pages/StatisticsIndex/staff/StaffStatistics.tsx (hunk 4 above)
Comment (bivanalhar) [reply client/app/bundles/course/statistics/pages/StatisticsIndex/staff/StaffStatistics.tsx:14]: done..

I was following the way `UserManagementTabs` is being used, since this component is my reference while implementing this StatisticsTabs. But after looking again, apparently this `UserManagementTabs` component is about to be refactored..

File: client/app/api/course/Statistics/CourseStatistics.js (hunk 2 above)
Comment (bivanalhar) [reply client/app/api/course/Statistics/CourseStatistics.js:45]: done

File: client/app/bundles/course/statistics/pages/StatisticsIndex/course/StudentPerformanceTable.tsx (hunk 1 above)
Comment (bivanalhar) [reply client/app/bundles/course/statistics/pages/StatisticsIndex/course/StudentPerformanceTable.tsx:136]: done

File: client/app/bundles/course/statistics/handles.ts (hunk 3 above)
Comment (ekowidianto) [reply client/app/bundles/course/statistics/handles.ts:22]: Do you know what this is even for?

Comment (bivanalhar) [reply client/app/bundles/course/statistics/handles.ts:22]: understand that this is for breadcrumbs, apologies for the oversight.
Have been reverted and fixed accordingly following the new conventions for course statistics routing
